### DISCUSSION TITLE
Best way of handling Websocket "rooms" and responding to multiple clients

### DISCUSSION THREAD

**Discussion by the-kwisatz-haderach** [marblejs/marble#308] [Q&A] [answered]
I'm working on a turn-based game where I want to broadcast server responses to particular subsets of clients, being part of the same game session or "room". I've just created a simple setup following the docs, but I'm stuck on how/where to handle this.

Unless I'm missing something, I'll proceed with something along the following lines:
When a client joins a game, I could store the `WebSocketClientConnection` from the effect context, keyed on the gameId. Then instead of responding with another event to the particular client which updates the game, I'll loop over all of the stored client connections for that gameId and run the `client.sendResponse` with a response.

My simple server bootstrapping for reference:

```
webSocketListener({
  middlewares: [],
  effects: gameEffects, // create initial game state, update game state.
  error$: errorHandler$
})

const webSocketServer = createWebSocketServer({
  options: {
    port: WEBSOCKET_PORT,
    host: HOST
  },
  listener: webSocketListener
})
```

**Comment by JozefFlakus** [comment, marked as answer]
Hi, finally the first topic in the discussions channel! 🙃 

The framework by design doesn't handle any form of channel/room support. It is up to the developer to implement this or reuse an existing solution (available somewhere? 🙈 ). If I would like to face this problem I would start with some code-responsibility detection.

1. You can reuse the DI mechanism and define a dependency/reader that will manage all incoming/outgoing and existing connections, eg. it doesn't mean that you have to store them explicitly - it is already managed by an instance of `WebSocketServerConnection` which is just an extension of `ws` `WebSocket.Server` interface. See [docs](https://github.com/websockets/ws/blob/master/doc/ws.md#serverclients). Each stored client has its own `id` property, added by the framework upon successful connection, that you can reuse. The trickiest thing that you have to handle here is the reader registration - currently the websocket server instance is returned as a result of  `webSocketServer` evaluation so you have to find a way how it can be accessible inside of the reader registered previously.

1. Room manager can also be modeled via context reader. It can reuse the previously mentioned custom reader

1. Joining a room can be done via the event handler that you previously mentioned - it will reuse the room manager from previous point.

Why it is better to handle connections via `WebSocketServerConnection` interface? It is safer - you don't have to handle and manage incoming and outgoing connections by yourself + you don't have to bother with potential memory leaks.

Maybe I can monkey-patch the `client` interface by adding the `server` attribute to it but I'm not sure if it won't cause any issues. 🤔 
[/packages/websockets/src/server/websocket.server.ts#L70](https://github.com/marblejs/marble/blob/master/packages/websockets/src/server/websocket.server.ts#L70)

**Reply by the-kwisatz-haderach** [reply]
Awesome, thanks for a thorough (and quick) response! I'll try to solve this following your pointers.
I'm a total RxJs newbie but this library is really cool. Keep it up. 👏
